src/predictive-text-input/predictive-text-input.tsx

```diff
@@ -0,0 +1,194 @@
+import debounce from "lodash/debounce";
+import React, { useCallback, useEffect, useState } from "react";
+import { Input } from "../input";
+import { DropdownList } from "../shared/dropdown-list/dropdown-list";
+import { DropdownWrapper } from "../shared/dropdown-wrapper";
+import {
+    Divider,
+    SelectorDiv,
+} from "../shared/dropdown-wrapper/dropdown-wrapper.styles";
+import { PredictiveTextInputProps } from "./types";
+
+export const PredictiveTextInput = <T, V>({
+    className,
+    "data-testid": testId,
+    selectedOption,
+    minimumCharacters = 3,
+    fetchOptions,
+    placeholder = "Enter here...",
+    readOnly = false,
+    disabled = false,
+    error,
+    valueExtractor,
+    listExtractor,
+    displayValueExtractor,
+    onSelectOption,
+}: PredictiveTextInputProps<T, V>) => {
+    // =============================================================================
+    // CONST, STATE
+    // =============================================================================
+    const inputValue = selectedOption && displayValueExtractor(selectedOption);
+    const [input, setInput] = useState<string>(inputValue);
+    const [searchedInput, setSearchedInput] = useState<string>(inputValue);
+    const [options, setOptions] = useState<T[]>([]);
+    const [isLoading, setIsLoading] = useState<boolean>(true);
+    const [prevOptionSelected, setPrevOptionSelected] =
+        useState<T>(selectedOption);
+
+    // =============================================================================
+    // DEBOUNCE FUNCTIONS
+    // =============================================================================
+
+    const handleFetchOptions = async (input: string) => {
+        setIsLoading(true);
+        const fetchedOptions = await fetchOptions(input);
+        setSearchedInput(input);
+        setOptions(fetchedOptions);
+        setIsLoading(false);
+    };
+
+    const fetchOptionsDebounced = useCallback(
+        debounce((input) => handleFetchOptions(input), 500, {
+            leading: false,
+            trailing: true,
+        }),
+        []
+    );
+
+    // =============================================================================
+    // EFFECTS
+    // =============================================================================
+
+    useEffect(() => {
+        if (
+            input &&
+            input.length >= minimumCharacters &&
+            input !== searchedInput
+        ) {
+            fetchOptionsDebounced(input);
+        }
+
+        if (input === "") {
+            setPrevOptionSelected(undefined);
+        }
+    }, [input]);
+
+    useEffect(() => {
```

Comment:
I think this can be a general effect to sync `prevOptionSelected` and `input` to `selectedOption`